src/mixing_fab.cc

```diff
@@ -0,0 +1,196 @@
+#include "mixing_fab.h"
+#include <sstream>
+
+using cyclus::Material;
+using cyclus::Composition;
+using cyclus::toolkit::ResBuf;
+
+using pyne::simple_xs;
+
+namespace cycamore {
+  
+  
+MixingFab::MixingFab(cyclus::Context* ctx)
+: cyclus::Facility(ctx), throughput(0) {
+  cyclus::Warn<cyclus::EXPERIMENTAL_WARNING>(
+     "the MixingFab archetype "
+     "is experimental");
+}
+
+cyclus::Inventories MixingFab::SnapshotInv() {
+  cyclus::Inventories invs;
+  
+  // these inventory names are intentionally convoluted so as to not clash
+  // with the user-specified stream commods that are used as the MixingFab
+  // streams inventory names.
+  invs["output-inv-name"] = output.PopNRes(output.count());
+  output.Push(invs["output-inv-name"]);
+  
+  std::map<std::string, ResBuf<Material> >::iterator it;
+  for (it = streambufs.begin(); it != streambufs.end(); ++it) {
+    invs[it->first] = it->second.PopNRes(it->second.count());
+    it->second.Push(invs[it->first]);
+  }
+  return invs;
+}
+
+void MixingFab::InitInv(cyclus::Inventories& inv) {
+  inv["output-inv-name"] = output.PopNRes(output.count());
+  output.Push(inv["output-inv-name"]);
+  
+  cyclus::Inventories::iterator it;
+  for (it = inv.begin(); it != inv.end(); ++it) {
+    streambufs[it->first].Push(it->second);
+  }
+}
+
+
+void MixingFab::EnterNotify() {
+  cyclus::Facility::EnterNotify();
+
+  if(fill_commod_prefs.size() == 0){
+    for (int i = 0; i < in_commods.size(); i++) {
+      fill_commod_prefs.push_back(1);
+    }
+  } else if (fill_commod_prefs.size() != in_commods.size()) {
+    std::stringstream ss;
+    ss << "prototype '" << prototype() << "' has " << mixing_ratio.size()
+    << " commodity preferences values, expected " << in_commods.size();
+    throw cyclus::ValidationError(ss.str());
+  }
+    
+  
+  if (mixing_ratio.size() == 0) {
+    for (int i = 0; i < in_commods.size(); i++) {
+      mixing_ratio.push_back(1./in_commods.size());
+    }
+  
+  } else if (mixing_ratio.size() != in_commods.size()) {
+    std::stringstream ss;
+    ss << "prototype '" << prototype() << "' has " << mixing_ratio.size()
+    << " commodity fraction values, expected " << in_commods.size();
+    throw cyclus::ValidationError(ss.str());
+  
+  } else {
+    double frac_sum = 0;
+    for( int i = 0; i < mixing_ratio.size(); i++)
+      frac_sum += mixing_ratio[i];
+    
+    if(frac_sum != 1.) {
```

Comment:
Also, it is bad to do a `!=` or a `==` comparison for floats.  This should be `>=`